### PR TITLE
Explorer: add more information to stack trace

### PR DESCRIPTION
**Pull request by prabhatexit0**
Adds **action kind** and **source location** for stack trace.
Example:
```
(+) stack-push: <action> (<source location>)
(+) stack-pop:  <action> (<source location>)
```
This PR also updates the `Action::Print(...)` method, following is the generalized view of `Action::Print(...)`.
```
ActionKind pos: <pos_count> `<related_code>` results: [<collected_results>]  scope: [<scope>]
```

### CODE REVIEW

File: explorer/trace_testdata/full_trace.carbon [hunk 1]
```diff
@@ -93,18 +93,18 @@ fn Main() -> i32 {
 // CHECK:STDOUT: checking BindingPattern n: i32
 // CHECK:STDOUT: checking ExpressionPattern i32
 // CHECK:STDOUT: checking IntTypeLiteral i32
-// CHECK:STDOUT: (+) stack-push: i32 .0.
-// CHECK:STDOUT: (+) stack-push: i32 .0.
+// CHECK:STDOUT: (+) stack-push: pos-0 ValueExpressionAction `i32` (full_trace.carbon:11)
+// CHECK:STDOUT: (+) stack-push: pos-0 ExpressionAction `i32` (full_trace.carbon:11)
 // CHECK:STDOUT: --- step exp i32 .0. (full_trace.carbon:11) --->
```
Comment (geoffromer) [inline explorer/trace_testdata/full_trace.carbon]: It seems like the code generating this line should be updated to use `Action::Print` for consistency. Note that you'll also need to update `explorer/README.md` to reflect the new trace format.

File: explorer/trace_testdata/full_trace.carbon [hunk 2]
```diff
@@ -93,18 +93,18 @@ fn Main() -> i32 {
 // CHECK:STDOUT: checking BindingPattern n: i32
 // CHECK:STDOUT: checking ExpressionPattern i32
 // CHECK:STDOUT: checking IntTypeLiteral i32
-// CHECK:STDOUT: (+) stack-push: i32 .0.
-// CHECK:STDOUT: (+) stack-push: i32 .0.
+// CHECK:STDOUT: (+) stack-push: pos-0 ValueExpressionAction `i32` (full_trace.carbon:11)
```
Comment (geoffromer) [inline explorer/trace_testdata/full_trace.carbon]: It seems really surprising to me to print `pos` before the action kind. `pos` (and the rest of the information) isn't meaningful until you know the kind, so it really seems like the kind should go first.

Comment (geoffromer) [inline explorer/trace_testdata/full_trace.carbon]: `pos-0` isn't a very intuitive notation to me. How about something like `pos: 0` for consistency with things like `results: ...`

File: explorer/trace_testdata/full_trace.carbon [hunk 3]
```diff
@@ -197,153 +197,123 @@ fn Main() -> i32 {
...
-// CHECK:STDOUT: (-) stack-pop: Main .1. {{[[][[]}}fun<Main>]]
+// CHECK:STDOUT: (-) stack-pop:  pos-0 ExpressionAction `Main` (<Main()>:0)
+// CHECK:STDOUT: (-) stack-pop:  pos-1 ValueExpressionAction `Main` results: [`fun<Main>`]  (<Main()>:0)
 // CHECK:STDOUT: --- step exp Main() .1. (<Main()>:0) --->
 // CHECK:STDOUT: calling function: fun<Main>
 // CHECK:STDOUT: match pattern ()
 // CHECK:STDOUT: from value expression with value ()
-// CHECK:STDOUT: (+) stack-push: .0. {}
-// CHECK:STDOUT: (+) stack-push: {var x: i32 = N.Foo(0);return x;} .0.
+// CHECK:STDOUT: (+) stack-push: pos-0 ScopeAction `` scope: [] (None)
```
Comment (geoffromer) [inline explorer/trace_testdata/full_trace.carbon]: The "``" seems unnecessary here.

File: explorer/trace_testdata/full_trace.carbon [hunk 4]
```diff
@@ -197,153 +197,123 @@ fn Main() -> i32 {
...
-// CHECK:STDOUT: (-) stack-pop: {return (n + 1);} .1. {}
-// CHECK:STDOUT: (-) stack-pop: .0. {n: i32: 0}
-// CHECK:STDOUT: (+) stack-push: clean up.0. {n: i32: 0}
-// CHECK:STDOUT: (+) stack-push: clean up.0. {}
-// CHECK:STDOUT: (-) stack-pop: clean up.0. {}
-// CHECK:STDOUT: (-) stack-pop: clean up.0. {n: i32: 0}
+// CHECK:STDOUT: (-) stack-pop:  pos-1 StatementAction `return (n + 1);` results: [`1`]  (full_trace.carbon:12)
+// CHECK:STDOUT: (-) stack-pop:  pos-1 StatementAction `{return (n + 1);}` scope: [] (full_trace.carbon:13)
+// CHECK:STDOUT: (-) stack-pop:  pos-0 ScopeAction `` scope: [`n: i32`: `0`] (None)
+// CHECK:STDOUT: (+) stack-push: pos-0 CleanUpAction `clean up` scope: [`n: i32`: `0`] (stack cleanup:1)
```
Comment (geoffromer) [inline explorer/trace_testdata/full_trace.carbon]: "\`clean up\`" seems redundant, now that the output already tells me it's a `CleanUpAction`.

File: explorer/trace_testdata/full_trace.carbon [hunk 5]
```diff
@@ -197,153 +197,123 @@ fn Main() -> i32 {
...
-// CHECK:STDOUT: (+) stack-push: clean up.0. {x: i32: lval<Allocation(1)>}
+// CHECK:STDOUT: (-) stack-pop:  pos-1 StatementAction `return x;` results: [`1`]  (full_trace.carbon:17)
+// CHECK:STDOUT: (-) stack-pop:  pos-2 StatementAction `{var x: i32 = N.Foo(0);return x;}` scope: [`x: i32`: `lval<Allocation(1)>`] (full_trace.carbon:18)
+// CHECK:STDOUT: (-) stack-pop:  pos-0 ScopeAction `` scope: [] (None)
+// CHECK:STDOUT: (+) stack-push: pos-0 CleanUpAction `clean up` scope: [] (stack cleanup:1)
+// CHECK:STDOUT: (+) stack-push: pos-0 CleanUpAction `clean up` scope: [`x: i32`: `lval<Allocation(1)>`] (stack cleanup:1)
 // CHECK:STDOUT: +++ memory-read: #1 `1`
-// CHECK:STDOUT: (+) stack-push: destroy.0.
-// CHECK:STDOUT: (-) stack-pop: destroy.0.
+// CHECK:STDOUT: (+) stack-push: pos-0 DestroyAction `destroy` (None)
```
Comment (geoffromer) [inline explorer/trace_testdata/full_trace.carbon]: "\`destroy\`" seems redundant, like "\`clean up\`"

File: explorer/trace_testdata/full_trace.carbon (hunk 1 above)
Comment (prabhatexit0) [reply explorer/trace_testdata/full_trace.carbon]: I was thinking of updating the interpreter related trace output in a separate PR.
Yes, I'll update the documentation for the new format. Thanks for pointing it out!

File: explorer/trace_testdata/full_trace.carbon (hunk 2 above)
Comment (prabhatexit0) [reply explorer/trace_testdata/full_trace.carbon]: Oh okay, I'll move it after `ActionKind` in the next commit.

Comment (prabhatexit0) [reply explorer/trace_testdata/full_trace.carbon]: Yes I think its reasonable to go with `pos: <num>` for consistency.

File: explorer/trace_testdata/full_trace.carbon (hunk 3 above)
Comment (prabhatexit0) [reply explorer/trace_testdata/full_trace.carbon]: Thanks for pointing this out!

File: explorer/README.md [hunk 6]
```diff
@@ -200,33 +200,71 @@ The following options can be passed as a comma-separated list to the
...
+memory is updated as objects are allocated and deallocated, and the stack is
+updated as actions are performed. The state of the program can be used to track
+the progress of the program and to debug the program.
+
+#### Memory
 
 The memory is a mapping of addresses to values. The memory is used to represent
-both heap-allocated objects and also mutable parts of the procedure call stack,
-for example, for local variables. When an address is deallocated, it stays in
-memory but `!!` is printed before its value.
```
Comment (geoffromer) [inline explorer/README.md:189]: If this part is still true, we should still mention it in the README somewhere.

File: explorer/trace_testdata/full_trace.carbon (hunk 1 above)
Comment (geoffromer) [reply explorer/trace_testdata/full_trace.carbon]: I'd prefer not to introduce an inconsistency, even temporarily. Would updating the `--- step` output in this PR create other inconsistencies?

**Comment by prabhatexit0** [comment]
> I'd prefer not to introduce an inconsistency, even temporarily. Would updating the --- step output in this PR create other inconsistencies?

@geoffromer Yes it makes sense, I've modified the step trace to use new action print method. Thanks!

File: explorer/README.md (hunk 6 above)
Comment (prabhatexit0) [reply explorer/README.md:189]: I'm not able to figure it out if this should be added or not.
For printing memory deallocation we have 
```
(+) memory-dealloc: #<allocation_index> `value`
```
Please let me know if I'm missing something.

Comment (geoffromer) [reply explorer/README.md:189]: This is about when the trace prints the contents of the heap. It looks like heap printing was moved out of trace logging at some point, although the code to do it is still there. So I guess this is OK to remove.

**Review by geoffromer — APPROVED** [review]

**Pull request merged**